### PR TITLE
[AMDAIEFuseFillIntoForall] Handle case where fill output is not sliced 

### PR DESCRIPTION
**Pull request by newling**
For 2x2 or 4x4 tiling the chain of ops after the fill looks like

```
 %9 = linalg.fill ins(%cst : f32) ...
 ... 
 %12 = scf.forall (%arg3, %arg4) in (2, 2) shared_outs(%arg5 = %9)  ...
 ...
 %extracted_slice_19 = tensor.extract_slice %arg5 ... 
 ...
 %19 = linalg.generic  ... outs(%extracted_slice_19 ... )
 ```
 
 i.e. the filled value enters an extract_slice inside the scf.forall. But for 1x1 tiling, it looks like 
 
 
 ```
  %9 = linalg.fill ins(%cst : f32)
  ...
  %14 = scf.forall (%arg3, %arg4) in (1, 1) shared_outs(%arg5 = %9)
  ...
  %19 = linalg.generic  ... outs(%arg5... )
```

i.e. there is no intermediate extract_slice. 

Before this PR, the logic was hardcoded to look for an extrac_slice, this PR relaxes this. 

Before this PR, 1x1 tiling hits https://github.com/nod-ai/iree-amd-aie/blob/f5ab91ebc3eb9fd7ffd8e53bd53aed434e893c5f/compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp#L71

After this PR, the compilation progresses further (fails much later in objectfifo pipeline, unrelated to this).

### CODE REVIEW

**Comment by yzhang93** [comment]
I think in applications when we see loops like

`%14 = scf.forall (%arg3, %arg4) in (1, 1) shared_outs(%arg5 = %9)`

The forall should be canonicalized away and we don't have to fuse the fill op into the loop.

**Comment by newling** [comment]
The canonicalizer doesn't ever remove scf.forall ops which thread/block ids afaik

**Comment by yzhang93** [comment]
> > However, I can see why this is needed here. We want to keep the thread mapping attribute?
> 
> The canonicalizer doesn't ever remove scf.forall ops with thread/block ids afaik. Which is, as I think you're saying, exactly what we want.

We can always modify the canonicalize patterns to include such case. But the real question is do we want such loop to be canonicalized away? In our current pipeline, a lot of passes have dependency on thread mapping attribute.

**Comment by yzhang93** [comment]
> I agree that if it were canonicalized away, we'd have many issues in our passes. That is why this PR doesn't try to eliminate the scf.forall.

The reason I'm discussing this here is that when we initially created these individual passes, the general principle was to reuse as many upstream functions as possible, such as `scf::tileAndFuseProducerOfSlice` in this pass (it is also used in FusePackIntoLoop). I think the upstream function also doesn't handle the case when loop count is 1, since the loop should have already been canonicalized. Is there anything we can do to keep using this upstream function while adding support for the additional case? CC: @Abhishek-Varma

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp [hunk 1]
```diff
@@ -32,58 +33,67 @@ void AMDAIEFuseFillIntoForallPass::runOnOperation() {
...
-  // Materialize the slice of the producer in place.
-  std::optional<scf::SCFFuseProducerOfSliceResult> fusedProducer =
-      scf::tileAndFuseProducerOfSlice(rewriter, sliceOpToTile,
-                                      MutableArrayRef(&loops, 1));
-  if (!fusedProducer) {
-    funcOp->emitOpError("Failed to fuse fill op into forall loop.");
-    return signalPassFailure();
-  }
+  // Do not use the result of the old fill.
+  rewriter.replaceAllUsesWith(fillOp.getResults()[0], fillOp.getOutputs()[0]);
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp:96]: Can't this be included in the above replaceUsesWithIf?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/fuse_fill_into_forall.mlir [hunk 2]
```diff
@@ -29,3 +29,43 @@ func.func @fuse_fill_into_forall(%arg0: tensor<1x4x16x64xi8>, %arg1 : tensor<4x1
...
+  %1 = tensor.empty() : tensor<8xi8>
+
+  // check that the operand of scf.forall is not the filled tensor, because the
+  // fill will take place inside the scf.forall.
+  // CHECK: %[[FORALL:.*]] = scf.forall (%[[ARG1:.*]]) in (1)
+  // CHECK-SAME: shared_outs(%[[ARG2:.*]] = %[[FUNCARG]])
+
+  // check for the new fill
+  // CHECK: %[[NEWFILL:.*]] = linalg.fill
+  // CHECK-SAME: outs(%[[ARG2]] : tensor<8xi8>) -> tensor<8xi8>
```
Comment (yzhang93) [inline compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/fuse_fill_into_forall.mlir:51]: I feel it's not readable if you mixed `CHECK` with the comments. I'd prefer you put all the comments above and keep `//CHECK` section compact.

**Review by yzhang93 — CHANGES_REQUESTED** [review]

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp [hunk 3]
```diff
@@ -32,58 +33,67 @@ void AMDAIEFuseFillIntoForallPass::runOnOperation() {
...
+  SmallVector<linalg::FillOp> fillOps;
+  funcOp->walk([&](linalg::FillOp fillOp) { fillOps.push_back(fillOp); });
+  if (fillOps.size() != 1) return;
+  linalg::FillOp fillOp = fillOps[0];
+  if (fillOp.getResults().size() != 1) return;
+
+  // Confirm that there is a unique user that is a forall, and match
+  // the block argument that is used by the fill op, or return.
+  ResultRange::use_range fillUses = fillOp->getUses();
+  if (std::distance(fillUses.begin(), fillUses.end()) != 1) return;
```
[the remark this answers is not in the document]
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp:46]: Doesn't look like that is a method on use_range

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp [hunk 4]
```diff
@@ -32,58 +33,67 @@ void AMDAIEFuseFillIntoForallPass::runOnOperation() {
...
 
-  // Search the first use by a scf::ForallOp user.
-  scf::ForallOp forallOp;
-  auto itProducerUses =
-      llvm::find_if(tileableProducer->getUses(), [&](OpOperand &use) {
-        forallOp = dyn_cast<scf::ForallOp>(use.getOwner());
-        return forallOp;
+  // In the case where there are no extract_slice ops, we manually create the
+  // fill at the beginning of the forall body.
+  assert(!extractSliceOp);
```
[the remark this answers is not in the document]
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp:78]: I find code with lots of asserts easier to read, but I've changed it to an if-else

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp (hunk 1 above)
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp]: No, different ops

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/fuse_fill_into_forall.mlir (hunk 2 above)
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/test/fuse_fill_into_forall.mlir]: Updated

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp]: It should work on ops. `fillOp->hasOneUse()` doesn't work?

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp [hunk 5]
```diff
@@ -32,58 +33,67 @@ void AMDAIEFuseFillIntoForallPass::runOnOperation() {
...
+  rewriter.setInsertionPointToStart(forallOp.getBody());
+  Value scalar = fillOp.value();
+  Location loc = fillOp.getLoc();
+  auto fusedFill = rewriter.create<linalg::FillOp>(loc, scalar, bbArg);
```
[the remark this answers is not in the document]
Comment (yzhang93) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp]: This change does not show up in the latest revision.

File: compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (newling) [reply compiler/plugins/target/AMD-AIE/iree-amd-aie/Transforms/AMDAIEFuseFillIntoForall.cpp]: Done

**Pull request merged**